helix-core/src/test/java/org/apache/helix/manager/zk/client/TestHelixZkClient.java

```diff
@@ -163,6 +163,8 @@ public void handleDataDeleted(String s) {
       // expected to be here.
     }
 
+    // client B needs to re-install the data watch
+    sharedZkClientB.watchForData(TEST_PATH);
```

Comment:
Usually, I'm very cautious about changing the existing test case. This indicates that we changed the behavior. So after this change, the user will need to resubscribe the watch once it is deleted, right? Regardless of the Helix reserved node that has children change listeners to cover, how about the other zk path that we allow the user to subscribe listener? The listeners will be idle after the node has been removed once, right? This breaks the feature IMO.

Overall, I think we will need to make it optional. 1. It is fine that we change the behavior for the reserved patch with additional children change listener on the parent node. But for the other random path from user's requests, we should keep the old behavior.

Reply:
Your concern is valid! To think about it, actually, it is the reason why native zookeeper has 2 sets of path related watcher: exists type and data type. To make a distinction between two, IMO, we either create a new `subscribeXXXDataChange` method or add a parameter like `subscribeDataChange(xx, xx, boolean listenOnNonExistPath = True/False`. 

What's your opinion? In fact, at the current stage,  do we have users listen on random paths? If not, I think we're creating problems for ourselves.

Reply:
One way is keeping the existing subscription methods the same. But we created some private/protected methods for the HelixManager to use.
Alternatively, as you said, adding options to the method. It is not optimal but compared with adding more methods, I would prefer to keep it simple now. One improvement is that we extend the listeners so the one keeps existing after delete would be a different listener type internally. Then the callback knows what to do. It helps to keep the methods simple.

Reply:
Discussed offline and we agreed that the behavior change is necessary and the client needs to be aware of the change.